File: jena/src/main/scala/eu/ostrzyciel/jelly/convert/jena/JenaDecoderConverter.scala
package eu.ostrzyciel.jelly.convert.jena

import eu.ostrzyciel.jelly.core.ProtoDecoderConverter
import org.apache.jena.datatypes.RDFDatatype
import org.apache.jena.graph.*
import org.apache.jena.sparql.core.Quad

import scala.annotation.nowarn

final class JenaDecoderConverter extends ProtoDecoderConverter[Node, RDFDatatype, Triple, Quad]:
  override inline def makeSimpleLiteral(lex: String): Node = NodeFactory.createLiteralString(lex)

  override inline def makeLangLiteral(lex: String, lang: String): Node = NodeFactory.createLiteralLang(lex, lang)

  override inline def makeDtLiteral(lex: String, dt: RDFDatatype): Node = NodeFactory.createLiteralDT(lex, dt)

  override inline def makeDatatype(dt: String): RDFDatatype = NodeFactory.getType(dt)

  override inline def makeBlankNode(label: String): Node = NodeFactory.createBlankNode(label)

  override inline def makeIriNode(iri: String): Node = NodeFactory.createURI(iri)

  // TODO: change to .createTripleTerm in Jena 6. For Jena 5, we must keep using createTripleNode
  //   to maintain compatibility with 5.0.x–5.3.x.
  @nowarn("msg=deprecated")
  override inline def makeTripleNode(s: Node, p: Node, o: Node): Node = NodeFactory.createTripleNode(s, p, o)

  // See: https://github.com/apache/jena/issues/2578#issuecomment-2231749564
  override inline def makeDefaultGraphNode(): Node = Quad.defaultGraphNodeGenerated

  override inline def makeTriple(s: Node, p: Node, o: Node): Triple = Triple.create(s, p, o)

  override inline def makeQuad(s: Node, p: Node, o: Node, g: Node): Quad = Quad.create(g, s, p, o)

File: jena/src/main/java/eu/neverblink/jelly/convert/jena/JenaDecoderConverter.java
package eu.neverblink.jelly.convert.jena;

import eu.neverblink.jelly.core.ProtoDecoderConverter;
import eu.neverblink.jelly.core.utils.QuadMaker;
import eu.neverblink.jelly.core.utils.TripleMaker;
import org.apache.jena.datatypes.RDFDatatype;
import org.apache.jena.graph.Node;
import org.apache.jena.graph.NodeFactory;
import org.apache.jena.graph.Triple;
import org.apache.jena.sparql.core.Quad;

public final class JenaDecoderConverter
    implements ProtoDecoderConverter<Node, RDFDatatype>, TripleMaker<Node, Triple>, QuadMaker<Node, Quad>
{

    @Override
    public Node makeSimpleLiteral(String lex) {
        return NodeFactory.createLiteralString(lex);
    }

    @Override
    public Node makeLangLiteral(String lex, String lang) {
        return NodeFactory.createLiteralLang(lex, lang);
    }

    @Override
    public Node makeDtLiteral(String lex, RDFDatatype dt) {
        return NodeFactory.createLiteralDT(lex, dt);
    }

    @Override
    public RDFDatatype makeDatatype(String dt) {
        return NodeFactory.getType(dt);
    }

    @Override
    public Node makeBlankNode(String label) {
        return NodeFactory.createBlankNode(label);
    }

    @Override
    public Node makeIriNode(String iri) {
        return NodeFactory.createURI(iri);
    }

    @Override
    public Node makeTripleNode(Node s, Node p, Node o) {
        return JenaCompatHelper.getInstance().createTripleNode(s, p, o);
    }

    @Override
    public Node makeDefaultGraphNode() {
        return Quad.defaultGraphNodeGenerated;
    }

    @Override
    public Quad makeQuad(Node subject, Node predicate, Node object, Node graph) {
        return Quad.create(graph, subject, predicate, object);
    }

    @Override
    public Triple makeTriple(Node subject, Node predicate, Node object) {
        return Triple.create(subject, predicate, object);
    }
}
